torch/csrc/distributed/c10d/TCPStore.cpp

```diff
@@ -1236,20 +1259,43 @@ void TCPStore::wait(
 void TCPStore::doWait(
     c10::ArrayRef<std::string> keys,
     std::chrono::milliseconds timeout) {
-  // TODO: Should we revert to the original timeout at the end of the call?
-  client_->setTimeout(timeout);
+  {
+    detail::SendBuffer buffer(*client_, detail::QueryType::WAIT);
+    buffer.appendValue(keys.size());
+    for (const std::string& key : keys) {
+      buffer.appendString(key);
+    }
+    buffer.flush();
+  }
 
-  detail::SendBuffer buffer(*client_, detail::QueryType::WAIT);
-  buffer.appendValue(keys.size());
-  for (const std::string& key : keys) {
-    buffer.appendString(key);
+  detail::WaitResponseType response;
+  if (client_->receiveValueWithTimeout<detail::WaitResponseType>(
+          response, timeout)) {
+    if (response != detail::WaitResponseType::STOP_WAITING) {
+      TORCH_CHECK(false, "Stop_waiting response is expected");
+    }
+    return;
   }
-  buffer.flush();
+  // this is the cancel wait timeout, once here we expect the server to respond
+  // in a timely fashion
+  {
+    detail::SendBuffer buffer(*client_, detail::QueryType::CANCEL_WAIT);
+    buffer.flush();
+  }
+
+  response = client_->receiveValue<detail::WaitResponseType>();
+  // this can happen if the server responds before we cancel, just ignorte it
+  if (response != detail::WaitResponseType::WAIT_CANCELED) {
+    if (response != detail::WaitResponseType::STOP_WAITING) {
+      TORCH_CHECK(false, "Stop_waiting response is expected");
+    }
 
-  auto response = client_->receiveValue<detail::WaitResponseType>();
-  if (response != detail::WaitResponseType::STOP_WAITING) {
-    TORCH_CHECK(false, "Stop_waiting response is expected");
+    response = client_->receiveValue<detail::WaitResponseType>(); // ignore
```

Comment:
what is the purpose for this second check? In what case are there two STOP_WAITING messages sent?

Reply:
This covers the case where there's a race between client and server.

The server sends a STOP_WAITING response to the client at the same time the client sends a CANCEL_WAIT request.

In this case, we expect to get to receive two replies from the server .